server/src/tram_track_graph_transformer/tram_track_graph_transformer.py

```diff
@@ -0,0 +1,241 @@
+        an edge between two nodes in this graph, that means there is a path between
+        two permanent nodes in the `_tram_track_graph`, which doesn't contain any
+        other permanent nodes.
+        """
+
+        return nx.DiGraph(
+            (source_node, destination_node)
+            for source_node in self._permanent_nodes
+            for destination_node in self._reachable_permanent_nodes_dfs(source_node)
+        )
+
+    def _add_coordinates_from_data(self):
+        for node in self._minified_tram_track_graph.nodes:
+            node_id = int(node)
+            if node_id in self._node_coordinates_by_id:
+                lat, lon = self._node_coordinates_by_id[node_id]
+                self._minified_tram_track_graph.nodes[node]["lat"] = lat
+                self._minified_tram_track_graph.nodes[node]["lon"] = lon
+
+    def _tag_nodes_with_way_ids(self):
+        for way in self._ways:
+            way_id = way.id
+            for node in way.nodes:
+                node_id = node.id
+                if node_id in self._minified_tram_track_graph.nodes:
+
+                    if "ways" not in self._minified_tram_track_graph.nodes[node_id]:
+                        self._minified_tram_track_graph.nodes[node_id]["ways"] = []
+                    self._minified_tram_track_graph.nodes[node_id]["ways"].append(
+                        way_id
+                    )
+
+    @property
+    def permanent_node_ids(self):
+        return self._permanent_nodes.copy()
+
+    def densify_graph_by_max_distance(
+        self, max_distance_in_meters: float
+    ) -> nx.DiGraph:
+
+        if max_distance_in_meters <= 0:
+            raise ValueError("max_distance_in_meters must be a positive float value")
+
+        def interpolate_between_first_and_last_node(
+            path_with_coords, max_distance_in_meters
+        ):
+            latlon = [(lat, lon) for _, lat, lon in path_with_coords]
+
+            transformer = Transformer.from_crs("EPSG:4326", "EPSG:2180", always_xy=True)
+            inverse = Transformer.from_crs("EPSG:2180", "EPSG:4326", always_xy=True)
+
+            meter_coords = [transformer.transform(lon, lat) for lat, lon in latlon]
+            line = LineString(meter_coords)
+            length = line.length
+
+            if length <= max_distance_in_meters:
+                return [latlon[0], latlon[-1]]
+
+            num_segments = max(1, math.ceil(length / max_distance_in_meters))
+            distances = [i * (length / num_segments) for i in range(1, num_segments)]
+
+            interpolated_points = [line.interpolate(d) for d in distances]
+            interpolated_latlon = [
+                inverse.transform(p.x, p.y)[::-1] for p in interpolated_points
+            ]
+
+            return [latlon[0]] + interpolated_latlon + [latlon[-1]]
+
+        new_graph = nx.DiGraph()
+        new_id_counter = max(self._tram_track_graph.nodes) + 1
```

Comment:
I think we could use the `self._permanent_nodes` property here, it should provide better typing information:
```suggestion
        new_id_counter = max(self._permanent_nodes) + 1
```

Reply:
Good point. Changed.